rasterio/rpc.py

```diff
@@ -0,0 +1,68 @@
+from rasterio.compat import Iterable, UserDict, text_type
+class RPC(UserDict):
+    """Rational Polynomial Coefficients used to map (x, y, z) <-> (row, col).
+    
+    RPCs are stored in a dict-like structure with methods to serialize and deserialize
+    to/from GDAL metadata strings.
+    """
+    GDAL_RPC_KEYS = (
+        "ERR_BIAS",
+        "ERR_RAND",
+        "HEIGHT_OFF",
+        "HEIGHT_SCALE",
+        "LAT_OFF",
+        "LAT_SCALE",
+        "LINE_DEN_COEFF",
+        "LINE_NUM_COEFF",
+        "LINE_OFF",
+        "LINE_SCALE",
+        "LONG_OFF",
+        "LONG_SCALE",
+        "SAMP_DEN_COEFF",
+        "SAMP_NUM_COEFF",
+        "SAMP_OFF",
+        "SAMP_SCALE"
+    )
+
+    def __init__(self, data={}, **kwargs):
+        UserDict.__init__(self)
+        initdata = {}
+        initdata.update(data)
+        initdata.update(**kwargs)
+        initdata = dict(filter(lambda item: item[0] in self.GDAL_RPC_KEYS, initdata.items()))
+        self.data.update(initdata)
+
+    def __getitem__(self, key):
+        """Normal dict item getter."""
+        return self.data[key]
+
+    def __setitem__(self, key, val):
+        """Normal dict item setter."""
+        self.data[key] = val
+
+
+    def to_gdal(self):
+        """Serialize dict values to string."""
+        out = {}
+
+        for key, val in self.items():
+            if isinstance(val, Iterable):
+                out[key] = ' '.join(map(text_type, val))
+            else:
+                out[key] = text_type(val)
+
+        return out
+
+    @classmethod
+    def from_gdal(cls, rpcs):
+        """Deserialize dict values to float or list."""
+        out = {}
+
+        for key, val in rpcs.items():
+            vals = val.split()
+            if len(vals) > 1:
+                out[key] = [float(v) for v in vals]
+            else:
+                out[key] = float(vals[0])
+
+        return cls(out)
```

Comment:
@underchemist what would you think about implementing this class using attrs instead like we do for Window (and might do for GroundControlPoint in the future)? https://www.attrs.org/en/stable/examples.html. Doing so would prevent problems with dict key typos and keep out items that aren't part of the RPC model.

Reply:
Let me know what you think, I've done so now. 

- The repr is a bit clunky and looks something like 

```
In [10]: src.rpcs
Out[10]: RPC(err_bias=0.5, err_rand=0.5, height_off=89.0, height_scale=701.0, lat_off=49.2199, lat_scale=0.3093, line_den_coeff=[1.0, 0.0009222511757408093, 0.0009437608823165506, 0.0, 3.227813186168402e-07, 0.0, -1.209023819818124e-08, 1.847595567040099e-06, 5.799102000140301e-07, -4.460833665210641e-07, 0.0, 0.0, -6.034797295293836e-08, 0.0, 1.667569505660956e-07, -2.974442475526043e-08, 0.0, 0.0, 0.0, 0.0], line_num_coeff=[0.002001303029283497, -0.1860717345625879, -1.127856422682348, -4.632230551975493e-05, 0.001019881908582774, 5.673855447822687e-08, -8.698433522294479e-08, -0.00322675985508854, -0.001332756784778704, 0.0, 1.215249975522747e-08, -7.132377372126199e-07, 2.017197679474549e-06, 8.275158167048898e-08, -1.210585983708413e-06, -1.07835388717073e-06, 5.032973521799926e-07, 0.0, 1.716128319528072e-08, 0.0], line_off=5760.0, line_scale=5761.0, long_off=-123.176, long_scale=0.4534, samp_den_coeff=[1.0, 9.278262976396983e-05, 0.001781926782031641, -0.0006510570023623242, -0.0002216055849873611, 9.161290674286373e-07, 3.126587074446549e-06, 0.0003565361629769621, -2.582447705973245e-05, -9.228544337667984e-05, 5.29808516621947e-07, 1.025006482963347e-05, 7.478466127324454e-07, -1.692384939549647e-06, -1.124443674146492e-05, -1.793628425616464e-07, -2.791740249303018e-07, -2.820306656137878e-07, 4.593502012060843e-08, 1.035174961061441e-07], samp_num_coeff=[0.02202618393703774, 1.185886131197477, -0.2151710781539888, 0.03045218075295352, 0.002420581655336635, -4.398438360671764e-06, 5.871407208028941e-05, -0.02166676957828599, -0.0004180699044156175, -2.753492566174621e-05, -7.124639699900795e-06, -1.101195320211651e-05, -0.0001119110912711932, -0.000109618465373252, 0.0001183590823839227, 1.382552349641905e-05, 1.997075688106731e-05, 2.673528192748438e-05, -1.230207121465409e-06, -2.830467933081173e-06], samp_off=3724.0, samp_scale=3725.0)
```
I experimented a bit with concatenating the list attributes, although it doesn't really solve much
- In practice, if a user is defining RPCs from their own source it is easiest to instantiate as `rpcs = RPC(**dict_with_keys_matching_kwargs)`. I doubt this is a common usecase.
- More likely, a user may be copying RPCs from a template/source dataset in which case they can instantiate `rpcs = RPC.from_gdal(src.tags(ns='RPC'))` or better  yet

```
with rasterio.open('in.tif') as src:
    with rasterio.open('out.tif' mode='w', width=10, height=10, dtype='uint8', rpcs=src.rpcs) as dst:
        pass
```

- In writing tests, I noticed that GDAL doesn't actually care about `ERR_BIAS` and `ERR_RAND` metadata items and doesn't write them out. For now I've left those attributes in case they are relevant to user, but `to_gdal` will exclude those attributes if they aren't truth-y